langroid/agent/base.py

```diff
@@ -1738,18 +1749,31 @@ async def handle_tool_message_async(
         Asynch version of `handle_tool_message`. See there for details.
         """
         tool_name = tool.default_value("request")
-        handler_method = getattr(self, tool_name + "_async", None)
+        if hasattr(tool, "handler"):
+            handler_name = tool.default_value("handler")
+        else:
+            handler_name = tool_name
+        handler_method = getattr(self, handler_name + "_async", None)
         if handler_method is None:
             return self.handle_tool_message(tool, chat_doc=chat_doc)
         has_chat_doc_arg = (
             chat_doc is not None
             and "chat_doc" in inspect.signature(handler_method).parameters
         )
+        has_tool_name_arg = ("tool_name" in
+                             inspect.signature(handler_method).parameters)
         try:
-            if has_chat_doc_arg:
-                maybe_result = await handler_method(tool, chat_doc=chat_doc)
+            if has_tool_name_arg:
+                if has_chat_doc_arg:
+                    maybe_result = await handler_method(tool, tool_name=tool_name,
```

Comment:
`tool_name` is a direct function of `tool`, so it seems redundant to have this arg I think? See my comment in the test

Reply:
@pchalasani you are absolutely right! Sorry, my bad. Pushed a fixed version.